### PR TITLE
Fixes #4868: Technique "Monitor a file or directory content": Implement possibility to not monitor a file which is absent

### PR DESCRIPTION
**Pull request by ncharles**
https://www.rudder-project.org/redmine/issues/4868

### CODE REVIEW

**Comment by amousset** [comment]
Could you make two commits (using rudder-dev techniques) to make review easier?

File: maintained-techniques [hunk 1]
```diff
@@ -131,7 +131,7 @@ systemSettings/remoteAccess/sshKeyDistribution/2.0
 systemSettings/remoteAccess/sshKeyDistribution/3.0
 systemSettings/security/fileAlterationMonitoring/1.0
 systemSettings/security/fileAlterationMonitoring/2.0
-systemSettings/systemManagement/cronManagement/1.0
+systemSettings/security/fileAlterationMonitoring/2.1systemSettings/systemManagement/cronManagement/1.0
```
Comment (amousset) [inline maintained-techniques]: This looks broken

File: techniques/systemSettings/security/fileAlterationMonitoring/2.0/metadata.xml [hunk 2]
```diff
@@ -15,6 +15,7 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
 -->
 
 <TECHNIQUE name="Monitor a file or directory content">
+  <DEPRECATED>This technique version has been superseded by a new version. It will no longer be available in the next stable version of Rudder. Please upgrade to the latest version.</DEPRECATED>
```
Comment (amousset) [inline techniques/systemSettings/security/fileAlterationMonitoring/2.0/metadata.xml]: We should not deprecate it now, rudder-dev is wrong :)

File: maintained-techniques (hunk 1 above)
Comment (ncharles) [inline maintained-techniques]: yeah

File: techniques/systemSettings/security/fileAlterationMonitoring/2.1/metadata.xml [hunk 3]
```diff
@@ -42,6 +42,21 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
         <NAME>MONITOR_ENTITY_NAME</NAME>
         <DESCRIPTION>Path to the file or directory to monitor</DESCRIPTION>
       </INPUT>
+      <SELECT1>
+        <NAME>MONITOR_NOFILE_BEHAVIOR</NAME>
```
Comment (amousset) [inline techniques/systemSettings/security/fileAlterationMonitoring/2.1/metadata.xml]: I would prefer something boolean, like MONITOR_ERROR_NOFILE, or something like this.

**Comment by ncharles** [comment]
Commit modified

File: techniques/systemSettings/security/fileAlterationMonitoring/2.1/fileAlterationMonitoring.st [hunk 4]
```diff
@@ -0,0 +1,85 @@
+#####################################################################################
+# Copyright 2011 Normation SAS
```
Comment (amousset) [inline techniques/systemSettings/security/fileAlterationMonitoring/2.1/fileAlterationMonitoring.st]: 2017

File: techniques/systemSettings/security/fileAlterationMonitoring/2.1/fileAlterationMonitoring.st [hunk 5]
```diff
@@ -33,6 +33,10 @@ bundle agent file_alteration_monitor
       "isfile_${index}" expression => isplain("${monitor_entity[${index}][name]}");
       "isdir_${index}" expression => isdir("${monitor_entity[${index}][name]}");
 
+      # Should the file reports an error if the file does not exist ?
```
Comment (amousset) [inline techniques/systemSettings/security/fileAlterationMonitoring/2.1/fileAlterationMonitoring.st]: Should the technique report an error

File: techniques/systemSettings/security/fileAlterationMonitoring/2.1/fileAlterationMonitoring.st [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+                &TRACKINGKEY:{uuid |"monitor_entity[&i&][uuid]" string => "&uuid&";
+}&
+      "index" slist => getindices("monitor_entity");
+
+  classes:
+
+      "isfile_${index}" expression => isplain("${monitor_entity[${index}][name]}");
+      "isdir_${index}" expression => isdir("${monitor_entity[${index}][name]}");
+
+      # Should the file reports an error if the file does not exist ?
```
Comment (amousset) [inline techniques/systemSettings/security/fileAlterationMonitoring/2.1/fileAlterationMonitoring.st]: Should the technique report an error

**Comment by ncharles** [comment]
Commit modified

**Comment by amousset** [comment]
You can self-merge after fixied the year and comment.

**Comment by ncharles** [comment]
OK, merging this PR

**Comment by ncharles** [comment]
OK, merging this PR

**Comment by ncharles** [comment]
OK, merging this PR

**Comment by ncharles** [comment]
OK, merging this PR

**Comment by ncharles** [comment]
OK, merging this PR

**Pull request merged**